Nom Etudiant 1: Assia MASTOR
Nom Etudiant 2: Maimouna ABDERAHMAN

# TME 2 : Projet Detection de motifs - Hash Table

Nous allons développer des algorithmes pour chercher de motifs dans les données de ChipSeq de C. glabrata.
Pour commencer nous allons d'abord produire des données artificielles qui nous permettront de tester rapidement nos algorithmes. Ensuite nous allons chercher les motifs dans C. glabrata et analyser les résultats.

## Partie A : Recherche de pattern (motifs) identiques


1\. Faites une fonction pour générer aléatoirement des séquences artificielles, puis générer t=10 séquences de n=41 nucléotides chacune. Toutes les lettres peuvent être équiprobables pour la génération des séquences.

In [2]:
import random

nuc = ('A', 'C', 'G', 'T')

t=10 #nombre de sequences
n=41 #longueur des sequences
#Pour un simple teste
t=3 #nombre de sequences
n=8 #longueur des sequences

def generateRandomSequence(n:int, upper=False):
    """
    Genere une séquence nucléotidique aléatoire 
    entrée n : longueur de la sequence
    entrée upper : bool, si True les nucléotides seront minuscule, False majuscule
    sortie sequence : une séquence nucléotidique aléatoire 
    """
    sequence = ""
    
    for i in range (n):
        tmp=random.choice(nuc)
        sequence =sequence+tmp
        
    if upper== True:
        sequence= sequence.lower()
    return sequence

def generateRandomSequences(n:int, t:int):
    """
    Genere plusieurs séquences nucléotidiques aléatoires 
    entrée n : longueur des sequences
    entrée t : nombre de sequences
    sortie sequences : liste de sequences nucléotidiques aléatoires 
    """
    sequences = []
    
    for i in range (t):
        tmp=generateRandomSequence(n,False)
        sequences.append(tmp)
        
    return sequences
    
seqs = generateRandomSequences(n, t);

print (seqs)


['CAACACTC', 'TCCTATAG', 'GAGTCCGT']


2\. Nous allons maintenant implanter dans les séquences artificielles generés avant un motif de taille `k`=9 à des positions aléatoires (choisies uniformément le long de la séquence). 
Bonus : on considère une proportion ``f``=0.9 de séquences qui possèdent le motif.

In [3]:
# taille des motifs k = 9
k = 3 
# frequences d'occurences des motifs f = 0.9 
#f = 0.9 # Pour le bonnus
f = 1

import random


def insertMotif(sequence, motif, position):
    return sequence[:position] + motif + sequence[position:]

def implantMotifs(motifs:str,  sequences:list, f=1):
    """
    Insérer un motif dans des positions aléatoires des séquences
    entrée motifs : motif qui va être implanté dans les séquences
    entrée f : fréquence d'implantation si 1 toutes les sequences contiendra un motif
    entrée séquences : liste de sequences
    sortie modified_sequences: liste de séquences ayant le motif implanté
    """
   
    modified_sequences     = []
    
    for seq in (sequences): #Boucle sur la liste de sequences, pour chaque seq création d'une liste de position choisie aléatoirement 
        listepos=[]
        for i in range(f):
            listepos.append(random.randrange(len(seq)))
            
        seqtmp=seq
        #print(listepos) #État de la liste avant trie :
        listepos.sort(reverse = True) #Trie listepos par ordre croissant
        #print(listepos) #État de la liste aprés trie
        
        for pos in listepos: #Boucle sur les positions et insertion du motif en commancant par la fin de la seq cad les positions les plus grandes
            seqtmp = insertMotif(seqtmp, motifs, pos)
        modified_sequences.append(seqtmp)
        
    return modified_sequences

# Generate motifs
def generateMotifs(k, nM):
    return [generateRandomSequence(k, False) for rs in range(nM)]

#Test
#print (implantMotifs("lolll",["012345678910111213","bbbbbbbbbb","cccccccccc","dddddddddddd"],1))

motif = generateRandomSequence(3, True)
motif_implanted_seqs = implantMotifs(motif,  seqs, f)
print ("\nModified sequences:", motif_implanted_seqs)



Modified sequences: ['CAgccACACTC', 'TCCgccTATAG', 'GAGgccTCCGT']


3\. Faites une fonction pour chercher les $m$ motifs de taille $k$ les plus fréquents dans l'ensemble des séquences. Tester cette fonction sur un l'ensemble de séquences avec le motif implanté génère dans les questions précédentes. Faite aussi une fonction qu'affiche les $top$ motifs les plus fréquents. 

In [4]:
def searchMotifs(k:int, sequences:list):
    """
    Cherche les motifs de taille k dans un ensemble de séquences
    entrée k : taille du motif
    entrée séquences : liste de sequences
    sortie motifs: dictionnaire de motifs, clé=motif, valeur = fréquence d'observation
    >>>searchMotifs(3, ['TAAGTAA', 'TATAA', 'CTATC'])
    {'TAA': 3, 'AAG': 1, 'AGT': 1, 'GTA': 1, 'TAT': 2, 'ATA': 1, 'CTA': 1, 'ATC': 1}
    """
    motifs  = {};
    
    for seq in sequences:
        for imot in range(len(seq)-k+1): #Boucle sur indice de 0 à l'indice = taille de la sequence - taille du motif + 1 = indice du dernier motifs de taille k de la séquence
            
            tmp=""
            for n in range(k):
                tmp= tmp+seq[imot+n]
                
            if tmp not in motifs.keys(): 
                motifs[tmp]= 1
            else:
                val= motifs[tmp]
                val+=1
                motifs[tmp]=val
                
    return motifs

def getTopMotifs(motifs:dict(), top:int):
    """
    renvoyer les top motifs le plus frequent
    entrée motifs: dictionnaire de motifs, clé=motif, valeur = fréquence d'observation
    entrée top : les top plus frequent 
    sortie motifsfreq: dictionnaire contenant les top motifs les plus fréquents, clé=motif, valeur = fréquence d'observation
    >>>getTopMotifs({'TAA': 3, 'AAG': 1, 'AGT': 1, 'GTA': 1, 'TAT': 2, 'ATA': 1, 'CTA': 1, 'ATC': 1}, 2)
    {'TAA': 3, 'TAT': 2}
    """
    motifsfreq  = {}
    
    valDicoSorted= sorted(motifs.values(), reverse = True) # Permet liste des valeurs du dico triées par ordre croissant
    valDicS = sorted(list(set(valDicoSorted)), reverse = True) # Suppresion des doublons dans valDicoSorted
    tmp= valDicS[0]
    
    for tp in range(top):
        if tp < len(valDicS):
            tmpVal = valDicS[tp] 
            
            #Récupère tt item du dico admettant pour valeur tmpVal et donc motif dans le top et les ajoutes au dico résultat
            for k,v in motifs.items():
                if v == tmpVal :
                    motifsfreq[k]=v  
    
    return motifsfreq

#Test
motifsFound = searchMotifs(3, ['TAAGTAA', 'TATAA', 'CTATC'])
print (motifsFound)
#{'TAA': 3, 'AAG': 1, 'AGT': 1, 'GTA': 1, 'TAT': 2, 'ATA': 1, 'CTA': 1, 'ATC': 1}

topMotifs = getTopMotifs(motifsFound, 2)
print (topMotifs)
#{'TAA': 3, 'TAT': 2}

{'TAA': 3, 'AAG': 1, 'AGT': 1, 'GTA': 1, 'TAT': 2, 'ATA': 1, 'CTA': 1, 'ATC': 1}
{'TAA': 3, 'TAT': 2}


4\. Quell est la complexité de votre algorithme?

<font color='blue'>
La compléxité de notre algorithme dépend de la taille des entrées, la taille du dictionnaire (n) et des motifs (m) ainsi que top une constante conditionant notre boucle for.
 
- Notre boucle for a pour compléxité pire cas top*(n*m) (Car : top * compléxité len() + top * (n*m))
- (2* compléxité sorted()) + compléxité list() + compléxité set() + commpléxité boucle for = O(n^2)+ O(m)+O(m) + top*(n*m) = top*(n*m)+m+(n^2)
    
Finalement notre fonction getTopMotifs a pour compléxité pire cas : n(m^2)top + n^2 
    
Pour ce qui est de searchMotifs, len(sequences)*(Pour chaque seq dans sequences sa taille donc total somme taille des séquences dans la list de sequence)

</font>

5\. Certains motifs sont reverse complémentaires. Implanter des motifs reverse complémentaires dans les séquences déjà générés et en suite chercher ces motifs. Attention: vous devez réutiliser les fonctions développés précédemment, pas besoin de définir des nouvelles fonctions. 

In [5]:
import re

def reversecompl(seq:str):
    """Renvoie le brin complémentaire d’une séquence.
    entrée : sequence de nucléotides (brin sens)
    sortie : sequence de nucléotides (brin complementaire)
    >>> reversecompl('AACGTGGCA')
    'TGCCACGTT'
    """

    compl = {'A': 'T', 'C': 'G', 'G': 'C', 'T':'A','a': 't', 'c': 'g', 'g': 'c', 't':'a'}
    rev_compl = ""
    for i in seq:
        if i in compl.keys() :
            rev_compl = compl[i]+ rev_compl
        else :
            print("Erreur ce(s) nucléotide(s) n'existent pas : "+i)
            return
    return rev_compl

def reversecomplList(seq):
    res=[reversecompl(sq) for sq in seq]
    return res
    
#print(reversecompl('AACGTGGCA'))
#print(reversecomplList(motif_implanted_seqs))

def randomimplantReverCompl(seq):
    """Implémente pour chaque sequence de la liste un motif reverse complémentaire en minuscule de taille t aléatoire à un indice tirer aléatoirement entre 0 et taille sequence - 1
        Retourne une liste avec des mootif reverse compl implentés 
    """

    indRandom=0 #Sert pour créer de façon aléatoire le motif
    motifàreverseCompl=''
    t = random.randint(0, len(seq[0])) #Pour que la taille du motif reverse comple implanter ne soit plus grand que toute la séquence, car choix du motif de taille t dont le reverse comple est implémenter est fait en prenant aléatoirement u indice entre 0 et taille de la seéquence- taille motif fixé aléatoirement
    LseqRes=[]
    LseqResTMP=[]
    
    print("Sequences seq d'entrée : \n",seq)
    for sq in seq :
        indRandom = random.randint(0, len(sq)-t)
        #print("Indice début choix motif à reverse compl ",indRandom)
        motifàreverseCompl=sq[indRandom:indRandom+t+1] #Sélection du motif
        motifreverseCompl = reversecompl(motifàreverseCompl.lower()) #Motif en minuscule pour permettre d ele recherché par la suite
        #print("Motif Reverse complémentaire à partir de cet indice", motifreverseCompl)
        
        #Implantation du motif reverse compl
        LseqResTMP = implantMotifs(motifreverseCompl,[sq],1)
        #print(LseqResTMP)
        LseqRes=LseqRes+LseqResTMP #Ajout de la sequence avec motif implanter
        LseqResTMP=[]
    
    return LseqRes

listeAvecMotifreverse=randomimplantReverCompl(seqs)
print("Liste seqs avec des motif reverse compl implanter pour chaque séquence : \n",listeAvecMotifreverse)

def randomimplantReverCompl(seq):
    """Fontion qui recherche début et fin du motif reverse complémentaire implanté pour chaque séquence de la liste de sequence en entrée, retourne sous forme de liste de n-uplet
    """
    tmpMotif=[]
    listeRes=[] #Liste de liste de n'uplet, avec indicde de début et fin pour tout les motifs trouver de la sequence 
    
    for sq in seq :
        tmpMotif= re.findall('[atgc]+',sq) #Récupère tt les motifs reverse comple implanter en minuscule dans la sequence, cela dit par défaut randomimplantReverCompl implante un seul motif donc ds la liste tjr un seul motif pour la séquence considérer
        #print(tmpMotif)
        
        listeTmp=[]
        indicedeb=0
        for motif in tmpMotif: #Pour chaque motif dans la sequence récupère indice de début
            indicedeb=sq.find(motif)
            listeTmp.append(indicedeb)
            
        listeRes.append(listeTmp) #Ajout à la liste resultat
    
    return listeRes

res = randomimplantReverCompl(listeAvecMotifreverse)
print("Liste avec indice de début trouver pour chaque motif reverse compl implanter (en minuscule) ds la sequence : \n",res)
    

Sequences seq d'entrée : 
 ['CAACACTC', 'TCCTATAG', 'GAGTCCGT']
Liste seqs avec des motif reverse compl implanter pour chaque séquence : 
 ['CgagtgttAACACTC', 'TCctataggaCTATAG', 'GAGTCCGacggactcT']
Liste avec indice de début trouver pour chaque motif reverse compl implanter (en minuscule) ds la sequence : 
 [[1], [2], [7]]



## Partie B : Recherche de motifs identiques sur vos données

1\. Le fichier "C_glabrata_1000bp_only.fasta" contiens les séquences régulatrices pour tous les gènes de votre organisme. Nous avons pris les 1000bp en amont du codon start. Chercher les 50 motifs de taille 7 les plus fréquents dans les deux brim de votre génome.

In [6]:
import time

k=7
feq=10
top= 50
genome = "C_glabrata_1000bp_only.fasta"


def readFasta(fastaFileName):
    """
    Read a fasta file
    entrée fastaFileName: nom du fichier fasta
    sortie sequences: liste contenant toutes les sequences du fichier
    """
    sequence = []
    file = open(genome, "r")
    sequence = []
    for s in file:
        if s[0] != ">":
            sequence.append(s.strip().upper())
    return sequence

sequences = readFasta(genome)
brincompl = reversecomplList(sequences)
fusionListebrin1et2 = sequences + brincompl

#dicoMotifBr1= searchMotifs(k, sequences) #Dico de tt les motifs de taille 7 récupérer ds brim 1
#dicoMotifBr2= searchMotifs(k, brincompl)
#top50Motif7BR1 = getTopMotifs(dicoMotifBr1,50)
#top50Motif7BR2 = getTopMotifs(dicoMotifBr2,50)
dicoMotifBr1ET2= searchMotifs(k, fusionListebrin1et2) #Dico de tt les motifs de taille 7 récupérer
top50 = getTopMotifs(dicoMotifBr1ET2,50)

#print("Voici le top 50 des motifs de taille 7 les plus fréquents dans le brin 1 ^^\n",top50Motif7BR1)
#print("\n\nVoici le top 50 des motifs de taille 7 les plus fréquents dans le brin 2^^\n",top50Motif7BR2)
print("\n\nVoici le top 50 des motifs de taille 7 les plus fréquents dans les 2 brims du génome ^^\n",top50)

dicoMotifTaille8 =searchMotifs(8, fusionListebrin1et2)
print("\n\nBONUS, fréquence motif TGATTCAT :", dicoMotifTaille8['TGATTCAT']) 




Voici le top 50 des motifs de taille 7 les plus fréquents dans les 2 brims du génome ^^
 {'TTTTTTT': 16043, 'AAAAAAA': 16043, 'ATTTTTT': 5930, 'AAAAAAT': 5930, 'TATATAT': 5892, 'ATATATA': 5892, 'AATTTTT': 4926, 'AAAAATT': 4926, 'TTTTTTC': 4708, 'GAAAAAA': 4708, 'AAAATTT': 4613, 'AAATTTT': 4613, 'TTTCTTT': 4576, 'AAAGAAA': 4576, 'TATTTTT': 4480, 'AAAAATA': 4480, 'AAGAAAA': 4295, 'TTTTCTT': 4295, 'TTTTTCA': 4129, 'TGAAAAA': 4129, 'AAAATAA': 4074, 'TTATTTT': 4074, 'AAATAAA': 4044, 'TTTATTT': 4044, 'TTTTATT': 3999, 'AATAAAA': 3999, 'TTCTTTT': 3991, 'AAAAGAA': 3991, 'TATATAA': 3932, 'TTATATA': 3932, 'ATATATT': 3802, 'AATATAT': 3802, 'TTTTTTG': 3771, 'CAAAAAA': 3771, 'AAAAAAG': 3714, 'CTTTTTT': 3714, 'AGAAAAA': 3668, 'TTTTTCT': 3668, 'ATAAAAA': 3642, 'TTTTTAT': 3642, 'AAATATA': 3568, 'TATATTT': 3568, 'TATTATT': 3553, 'AATAATA': 3553, 'AAAATAT': 3483, 'ATATTTT': 3483, 'TAAAAAA': 3465, 'TTTTTTA': 3465, 'TTTATAT': 3369, 'ATATAAA': 3369, 'TCTTTTT': 3265, 'AAAAAGA': 3265, 'AATATAA': 3135, 'TTAT

2\. Quel sont les trois motifs de taille 7 les plus frequents? Pensez vous que ces motifs correspondent à de facteur de transcription connus? Justifier votre reponse.

<font color='blue'>

Les trois motifs de taille 7 les plus fréquents sont : 


{'TTTTTTT': 16043, 'AAAAAAA': 16043, 'ATTTTTT': 5930, 'AAAAAAT': 5930,'TATATAT': 5892, 'ATATATA': 5892}, 3 motif et leur reverse complémentaire.
    
Il s'agit de motif peu complexe présent à une grande fréquence et donc trés probablement connus, cela dit ils n'ont pas certainement une signification biologique. 

En revanche le motif 'ATATATA', fait penser à la TATA box ! Cette séquence d'ADN codée TATA se situe à environ 25 nucléotides en amont du premier nucléotide transcrit et sert en partie de lieu de reconnaissance à l'ARN polymérase chez les eucaryotes.

</font>

<font color='black'>
3\. BONNUS: Le motif TGATTCAT correspond au facteur de transcription Gcn4 qui est trés souvant trouvé dans le genome de levures. Est-ce que vous avez trouvé ce motif? Si oui avec quel frequence?

Oui nous avons trouver le motif et à une fréquence de 152 pour TGATTCAT (Taille 8)
    

</font>

In [7]:
motifGcn4 = ['TGATTCA']

def searchGivenMotif(motifsTrouve, motifSpecifique):
    """
    Cherche une liste de motifs specifiques dans un dictionaire de motifs trouvés
    entrée motifsTrouve : dictionnaire de motifs, clé=motif, valeur = fréquence d'observation
    entrée motifSpecifique: un motif specifique à chercher
    sortie frequence : la frenquence du motif
    sortie ranking : dans quelle position le motif a été trouvé
    """
    frequence = 0; ranking = 0
    
    for k,v in motifsTrouve.items(): 
        if k == motifSpecifique[0] :
            frequence = v
            
            #Recherche du classement 
            valDicoSorted= sorted(motifsTrouve.values()) # Permet liste des valeurs du dico triées par ordre croissant
            valDicS = sorted(list(set(valDicoSorted)), reverse = True) # Suppresion des doublons dans valDicoSorted
            
            ranking = valDicS.index(v)+1 #Nous récupérons l'indice qui fait ainsi référence au classement auquel nous ajoutons 1 car les indices commmencent à 0
            
    if frequence == 0 :
        print("Le Motif specifique recherché ne figure pas ds le dictionnaire")
    
    return frequence, ranking

Res = searchGivenMotif(dicoMotifBr1ET2, motifGcn4)
print("BONNUS : Fréquence et classement du motif motifGcn4 TGATTCA de taille 7 ",Res)

Res2 =searchGivenMotif(dicoMotifTaille8, ['TGATTCAT'])
print("BONNUS : Fréquence et classement du motif motifGcn4 TGATTCAT de taille 8",Res2)
print("L'énoncée de la question précédente nous donnait un motif de taille 8 pour motifGcn4, ce qui n'est pas le cas du bonnus")


BONNUS : Fréquence et classement du motif motifGcn4 TGATTCA de taille 7  (1185, 597)
BONNUS : Fréquence et classement du motif motifGcn4 TGATTCAT de taille 8 (298, 594)
L'énoncée de la question précédente nous donnait un motif de taille 8 pour motifGcn4, ce qui n'est pas le cas du bonnus


4\. Les motifs peu complexes (avec par exemple 5, 6 ou 7 fois la même lettre) sont courants dans les génomes, ils n'ont généralement pas de signification biologique. Dans le context de ce projet, vous pouvez eliminer aussi les motifis ayant deux lettre repetés, comme par exemple AGAGAGA. Faites une fonction pour éliminer les motifs peu complexes, _i. e._ qui ont au moins $m$ fois la même lettre ou qui ont deux lettre consecutive répétés. Dans quel position vous trouver Gcn4 apres enlever les motifs peu complexes?</font>. 

In [8]:
def removeLowComplexe(motifs):
    """
    Enleve les motifs peu complexe ayant 
    entrée motifs: dictionnaire de motifs, clé=motif, valeur = fréquence d'observation
    sortie motifsClean: dictionnaire de motifs sans les motifs peu complexe.
    """
    motifsClean = {}
    
    nb = 5 #Nombre maximum de lettre répétées
    motifsClean = {}
    
    for k,v in motifs.items(): #Boucle sur les couples key value du dico
        SetDesCaractèresComposantsLeMotif = set(k) #Suppression des doublons du motifs afin d'obtenir un set avec que les différents caractères qui le compose
        lenSet = len(SetDesCaractèresComposantsLeMotif)
        compteur = 0
        
        if lenSet >= 3:#si le motif k admet moins de 3 caractères différents il est ignorer car focément 2 meme lettres répétées ou la meme lettre est répétée plusieur fois à la suite 
            #Test pour savoir si un mm caractère du motif n'est pas présent plus de 5 fois à la suite
            nbOccurence= 0
            for car in SetDesCaractèresComposantsLeMotif :
                nbOccurence = k.count(car*5)
                
                #Si le car est présent 5 fois à la suite ou plus, on sort du for car le motif est peu complex sinon on incrémente le compteur qui si atteint taille lenSet de caractère du motif implique que le motif n'est pas peu complex et on l'ajoute au dictionnaire résultat
                if nbOccurence > 0 :
                    break
                
                else :
                    compteur  = compteur + 1
                
                #Si compteur = len du set on ajoute le motif au dico et on sort 
                if compteur == lenSet :
                    motifsClean[k]=v
    
    return motifsClean

#dicotop10= getTopMotifs(dicoMotif,10)
#print(dicotop10)

dicotest = {'AAAAAAA': 8385, 'TTTTTTT': 7658,'ATATATA': 3096,'ATATATAC': 3096, 'ATTTGTTT': 2970,'ATTTTTGTTT': 2970}
print("**** Test avant clean des motifs peu complex ",dicotest)
dicoclean = removeLowComplexe(dicotest)
print("**** Test aprés clean des motifs peu complex ",dicoclean)



**** Test avant clean des motifs peu complex  {'AAAAAAA': 8385, 'TTTTTTT': 7658, 'ATATATA': 3096, 'ATATATAC': 3096, 'ATTTGTTT': 2970, 'ATTTTTGTTT': 2970}
**** Test aprés clean des motifs peu complex  {'ATATATAC': 3096, 'ATTTGTTT': 2970}


5\. Le fichier "Sequence_by_Peaks_G*.fasta" contiens les regions de peak trouvé par ChipSeq, qui contient probablement le Facteur de Transcription que nous cherchons. Apliquer les fonctions precedents pour chercher les 3 motifs les plus fréquents dans les deux brim. Il faut bien evidement enlever les motifs peu complexe.

In [14]:
k=8
feq=5
top= 30

genome = "Sequence_by_Peaks_1.fasta"
sequence = readFasta(genome)
brincompl = reversecomplList(sequences)
fusionListebr1et2 = sequence + brincompl

print("Recherche de motifs de taille ",k," ...\n")
listeMotifs = searchMotifs(k, fusionListebr1et2)

print("Suppression des motifs peu complexe ... \n")
motifsFiltre = removeLowComplexe(listeMotifs)
#print(motifsFiltre)

print("Le TOP 3 des motifs les plus fréquents dans les deux brins sont (Certains motifs sont présents à de meme fréquence donc partagent le podium) : \n")
Top = getTopMotifs(motifsFiltre, 3)
print(Top)

#Test pr savoir fréquence des motifs présent avec certitude ds le fichier et de taille 8
#print("\nTest pour vérifier fréquences des motifs présent avec certitude dans le fichier Sequence_by_Peaks_1.fasta : \n")
#print(motifsFiltre['TTTGCTCA'])  #Sur Brin1
#print(motifsFiltre['CTCATTGG'])  #SurBrin1
#print(motifsFiltre['CTGATTGG'])  #SurBrinCompl
#print(motifsFiltre['TCCGCGGA'])  #SurBrinCompl
#print(motifsFiltre['TCCGTGGA'])  #SurBrinCompl
#print(motifsFiltre['CAGCAAAA'])  #SurBrin1
#print(motifsFiltre['CGGATCCG'])  #SurBrin1

#Motif de taille k=7
#print(motifsFiltre['TGCACCC'])

#Motif de taille k=10
#print(motifsFiltre['CCTTNAGGGT'])

Recherche de motifs de taille  8  ...

Suppression des motifs peu complexe ... 

Le TOP 3 des motifs les plus fréquents dans les deux brins sont (Certains motifs sont présents à de meme fréquence donc partagent le podium) : 

{'CTCATCGC': 893, 'GCGATGAG': 879, 'GCTCATCG': 837}


6\. Ulitser la base YEASTRACT : http://www.yeastract.com/formsearchbydnamotif.php pour chercher les motifs. 
Avez vous une indication pour le facteur de transcription impliqué ?


In [ ]:
Pour les motifs de taille 8 éventuellement présent, voici les fréquences auquel on les retrouve sur les 2 brins : 
'TTTGCTCA':165
'CTCATTGG':81
'CTGATTGG':75
'TCCGCGGA':9
'TCCGTGGA':48
'CAGCAAAA':192
'CGGATCCG':35
Aucun d'eux ne figure dans le top3 des motifs de taille 8 le splus fréquents : {'CTCATCGC': 893, 'GCGATGAG': 879, 'GCTCATCG': 837}
    
    
Motif Aft1 de taille 7 :
'TGCACCC' :130
Il ne figure pas non plus parmi le top 3 des motifs de taille 7 les plus fréquents : {'TTTTCAA': 1633, 'TTTCAAA': 1580, 'TTTGAAA': 1507}
      
    
Motif Zap de taille 10 :
Ce motif n'a pas du tout été trouvé et n'est donc associé à aucune valeur dans le dictionnaire associé aux différents motifs en valeur leur nb d'occurrence comptabilisée.
{'AGCTCATCGC': 483, 'GCGATGAGCT': 466, 'GCTCATCGCA': 419}
    
Hash Table est malheureusement un algorithme qui prend pas en compte les mutations dans le motif.
Finalement, nous pouvons conclure que notre motif est soit Aft1 soit l'un des motifs de taille 8... il est trop  tôt  pour conclure.